extensions/answer_summarizers/models.py

```diff
@@ -38,72 +38,84 @@
+    hashed into built-in collections.
     """
-    hashable_answer_values = [
-        _get_hashable_value(answer_dict['answer'])
-        for answer_dict in answer_dicts_list]
-    answer_frequencies = collections.Counter(hashable_answer_values)
-    return [
-        ([answer_dicts_list[idx]
-          for idx, val in enumerate(hashable_answer_values)
-          if val == hashable_answer][0], frequency)
-        for (hashable_answer, frequency) in answer_frequencies.most_common()]
 
+    @classmethod
+    def _get_hashable_value(cls, value):
+        """This function returns a hashable version of the input value.
+
+        It converts the built-in collections into their hashable counterparts
+        {list: tuple, set: frozenset, dict: (sorted tuple of pairs)}.
+        Additionally, their elements are converted to hashable values through
+        recursive calls. All other value types are assumed to already be
+        hashable.
+        """
+        if isinstance(value, list):
+            return tuple(cls._get_hashable_value(e) for e in value)
+        elif isinstance(value, set):
+            # Set elements are already hashable.
+            return frozenset(value)
+        elif isinstance(value, dict):
+            # Dict keys are already hashable, only values need converting.
+            return tuple(sorted(
+                (k, cls._get_hashable_value(v)) for k, v in value.iteritems()))
+        else:
+            # Any other type is assumed to already be hashable.
+            return value
+
+    def __init__(self, value, key=None):
+        """Wraps value into a hashable object. Bases the hash on key(value) when
+        key is provided, otherwise it will be based simply on value.
+        """
+        self.value = value
+        self.hash_value = (
+            self._get_hashable_value(value if key is None else key(value)))
+
+    def __hash__(self):
+        return hash(self.hash_value)
+
+    def __eq__(self, other):
+        if isinstance(other, _HashedValue):
+            return self.hash_value == other.hash_value
+        return False
 
-def _calculate_top_answer_frequencies(state_answers_dict, num_results):
+
+def _get_top_answers_by_frequency(answer_dicts, limit=None):
     """Computes the number of occurrences of each answer, keeping only the top
     num_results answers, and returns a list of dicts; each dict has keys
     'answer' and 'frequency'.
 
     This method is run from within the context of a MapReduce job.
     """
-    top_answer_counts_as_list_of_pairs = _count_answers(
-        state_answers_dict['submitted_answer_list'])[:num_results]
-
-    calculation_output = []
-    for item in top_answer_counts_as_list_of_pairs:
-        calculation_output.append({
-            'answer': item[0]['answer'],
-            'frequency': item[1],
-        })
+    hashed_answer_frequencies = collections.Counter(
+        _HashedValue(d, key=operator.itemgetter('answer')) for d in answer_dicts
```

Comment:
Can you do this without operator.itemgetter? E.g. using _HashedValue(answer_dict['answer'])?

Reply:
Done. Yeah... realized that it was totally unnecessary after taking another look!